### PR TITLE
[WIP] Very WIP test PR - Make kubeadm control plane joins more robust

### PR DESCRIPTION
**Pull request by randomvariable**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

File: bootstrap/kubeadm/internal/cloudinit/controlplane_join.go [hunk 1]
```diff
@@ -54,10 +86,23 @@ func NewJoinControlPlane(input *ControlPlaneJoinInput) ([]byte, error) {
 	// TODO: Consider validating that the correct certificates exist. It is different for external/stacked etcd
 	input.WriteFiles = input.Certificates.AsFiles()
 	input.WriteFiles = append(input.WriteFiles, input.AdditionalFiles...)
+	joinScript, err := generate("JoinControlplaneScript", controlPlaneJoinScript, input)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to generate user data for machine joining control plane")
```
Comment (neolit123) [inline bootstrap/kubeadm/internal/cloudinit/controlplane_join.go]: "Wrapf" -> "Wrap"

also better to: "JoinControlplaneScript" -> "JoinControlPlaneScript"

**Comment by chuckha** [comment]
CAPD does something pretty similar, except it uses the control loop for retries. 

If kubeadm fails, the container gets deleted and the reconcile loop starts over, recreating the container. 

That's less than ideal in a cloud environment since that would add many minutes for a handful of kubeadm failures, but I think it is an option that is more in line with infra-as-cattle rather than trying to modify one node over and over.

However, the time saved on cluster spin up might be worth it.

**Comment by neolit123** [comment]
from my discussion with @vincepri on slack:

> so i think a safer bet would be:
```
kubeadm reset -f # this should pass right way if there is no admin.conf, if there is an admin.conf that does not point to a valid cluster it might block for ~10 seconds
kubeadm join ...
^ if join fails retry the whole block
```

> from my experience this works in multi-CP testing...but if the removal of etcd member somehow messes the etcd cluster it would need etcd client intervention /shrug.

**Comment by neolit123** [comment]
but he mentioned that you've seen etcd quorum failures, so i leave the phase vs whole join to you.

**Comment by randomvariable** [comment]
We've resolved the majority of issues in https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/851 . Defensively adding retries in kubeadm for future consumption by Cluster API may be nice though.

File: bootstrap/kubeadm/internal/cloudinit/controlplane_join.go [hunk 2]
```diff
@@ -17,12 +17,44 @@ limitations under the License.
...
+  until $@; do
+    sleep 15
+    echo "Couldn't run command. Retrying..."
+  done
+}
+
+kubeadm join --config /tmp/kubeadm-controlplane-join-config.yaml {{.KubeadmVerbosity}}
+KUBEADM_RETURN=$?
+if [ ${KUBEADM_RETURN} -ne 0 ]; then
+  retry-until-working kubeadm join phase control-plane-join mark-control-plane --config /tmp/kubeadm-controlplane-join-config.yaml {{.KubeadmVerbosity}}
```
Comment (ncdc) [inline bootstrap/kubeadm/internal/cloudinit/controlplane_join.go]: If we get a nonzero exit, I don't believe there's any guarantee that it got "far enough" such that the only steps left are marking and updating status. If there was a wholesale failure (couldn't find/load the images needed, for example), the nonzero exit is a hard stop, vs. one where there was a blip talking to etcd...

**Pull request closed without merging**